Feature/166 mentor assign tags to challenge

mcallejo-10 opened the pull request:
# Challenge Form Tags Implementation

## Changes
- Implement dynamic tag loading based on selected programming language
- Add multi-select functionality for tags
- Update challenge creation to include selected tags
- Create mock data files for tags by programming language (temporary solution)
- Integrate tag selection with form submission

## Technical Details
- Created mock data structure for tags per language while endpoints are developed
- Tags are dynamically loaded when a language is selected
- Tags can be toggled on/off with a single click
- Selected tags are visually highlighted
- Tags are properly cleared when switching languages
- Tags are included in the final challenge object on submission

## Design Notes
While the implementation differs from the original Figma design, the current multi-select tag system was agreed upon with the Scrum Master to enhance UX by allowing users to select multiple tags simultaneously.

GuillemBo committed:
Change save token to cookies instead of localStorage

GuillemBo committed:
isLoggedIn auth function

GuillemBo committed:
clear code

mcallejo-10 committed:
feat: add HTML structure and styling for tags selection

mcallejo-10 committed:
feat: implement dynamic tags loading based on selected language

- Add language-specific tag mapping
- Load corresponding tags when language changes
- Reset selected tags on language switch
- Hide tags section when no language is selected

GuillemBo committed:
fix tests

mcallejo-10 committed:
feat: add selected tags to challenge object on form submission

mcallejo-10 committed:
refactor: remove unused arrow icons from assets

- Clean up unused UI elements
- Remove arrow icon assets that were not implemented in the final design

GuillemBo committed:
delete comment

GuillemBo commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -16,6 +16,9 @@ import { java } from '@codemirror/lang-java'
 import { python } from '@codemirror/lang-python'
 import { basicSetup } from 'codemirror'
 
+// Añadir el import al inicio del archivo
```
You can delete this comment

GuillemBo commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -136,7 +143,7 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy {
         this.languages = results || []
         console.log('Idiomas cargados:', this.languages) // Lista completa de idiomas
-        console.log(JSON.stringify(this.languages, null, 2))
+        // console.log(JSON.stringify(this.languages, null, 2))
```
You can delete this comment since is not doing anything

GuillemBo commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -136,7 +143,7 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy {
         // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
```
You can delete this comment

GuillemBo commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -168,6 +176,21 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy {
+    this.selectedTags = [] // Resetear tags seleccionados al cambiar de lenguaje
+  }
+
   // Validación del formulario (código original)
```
The code is already self-explanatory. You can delete the comment

GuillemBo commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -147,8 +154,9 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy {
+    this.loadTagsForLanguage(language)
     /* istanbul ignore next */
     // Actualiza CodeMirror con el nuevo lenguaje
     if (this.editor != null) {
```
You can clear the comments that have no use

GuillemBo commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -168,6 +176,21 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy {
+
+    const selectedLanguageTags = tagMap[language as keyof typeof tagMap]
+    this.currentTags = selectedLanguageTags?.results ?? []
+    this.selectedTags = [] // Resetear tags seleccionados al cambiar de lenguaje
```
you can clear the comment.

GuillemBo commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -147,8 +154,9 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy {
 
   // Nuevo método para manejar el cambio de lenguaje
   // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
-  onLanguageChange (language: string) {
+  onLanguageChange (language: string): void {
     this.challenge.language = language
+    this.loadTagsForLanguage(language)
     /* istanbul ignore next */
     // Actualiza CodeMirror con el nuevo lenguaje
```
you can clear this commented code

mcallejo-10 committed:
test: tag management test added

mcallejo-10 committed:
refactor: remove unnecessary comments from challenge form component

mcallejo-10 commented:
- Added comprehensive test suite for tag management functionality

- Removed redundant comments

GuillemBo reviewed (approved):
The code is functioning as expected, and the requested changes have been implemented.

GuillemBo committed:
undo unnecessary changes

ivilarop commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -205,4 +227,18 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy {
+    } else {
+      this.selectedTags.splice(index, 1)
+    }
+    console.log('Tags seleccionados:', this.selectedTags)
```
delete it

ivilarop commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -136,7 +143,6 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy {
         // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
         this.languages = results || []
         console.log('Idiomas cargados:', this.languages) // Lista completa de idiomas
```
delete it

ivilarop commented on src/app/models/create-challenge.interface.ts:
```diff
@@ -11,4 +11,6 @@ export interface CreateChallenge {
   level: 'EASY' | 'MEDIUM' | 'HARD'
   language: string
   solution: string
+  topic: string
```
what is it?

mcallejo-10 replied:
Update to match back call, Albert sent me this json: 

{
  "challengeTitle": "Test with tags",
  "description": "string",
  "level": "EASY",
  "language": "Java",
  "solution": "string",
  "topic": "ALL",
  "tags": [
    "Recursividad"
  ]
}

ivilarop replied:
Yes, I understand, but the concept of a topic doesn't apply to this change. I understand that we have to include it to regularize it with the backend, but it creates confusion because it seems like the topic functionality is being implemented. If it doesn't give an error without a topic, let's not include it.

ivilarop replied:
We've been discussing this with Albert, and we believe the best and most optimal option is to send the tag's UUIDs. This way, we don't force the backend to search the database (TagDocument) for the UUIDs before saving them to ChallengeDocument.

"tags": [
  "67e7e9938a177b9eec207e0a"  // the id_tag ​​of the "Recursion" document
]

mcallejo-10 replied:
I've removed the word "topic" from tags and verified that all functionality continues to work properly.

mcallejo-10 committed:
Remove comments and apply requested PR changes

ivilarop commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.spec.ts:
Does adding the tag id affect the tests?

mcallejo-10 replied:
tests updated

ivilarop commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -128,15 +135,12 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy {
       next: ({ results }) => {
         // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
         this.languages = results || []
         console.log('Idiomas cargados:', this.languages) // Lista completa de idiomas
```
must be deleted the console.log. Although it is a historic change, we can take advantage of it to clean up since it is a small change.

mcallejo-10 replied:
deleted

ivilarop commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.html:
```diff
@@ -144,6 +144,23 @@ <h4><span class="section-number">3</span> Información adicional</h4>
+                    type="button" 
+                    class="tag-button"
+                    [class.active]="isTagSelected(tag.tag_name)"
+                    (click)="onTagSelect(tag.tag_name)"
```
change for id tag

mcallejo-10 replied:
changed

ivilarop commented on src/app/modules/challenge/components/challenge-form/challenge-form.component.ts:
```diff
@@ -33,10 +36,14 @@ export class ChallengeFormComponent implements AfterViewInit, OnDestroy {
     language: '' as string,
-    solution: ''
+    solution: '',
+    topic: '',
```
As we discussed earlier, if removing the topic doesn't give an error, don't include it. If we want to include it in the future, we'll make the change.

mcallejo-10 replied:
I've removed the word "topic" from tags and verified that all functionality continues to work properly.

GuillemBo committed:
undo unnecessary changes 2

GuillemBo committed:
auth service changes

mcallejo-10 committed:
feat: implement tag selection with UUIDs in challenge form

- Update tag selection to use UUIDs instead of tag names
- Update tests to match new tag structure
- Fix tag selection functionality in form component
- Add tags array to challenge model

mcallejo-10 commented:
## Changes Made
- Updated tag selection functionality to use UUIDs instead of name
- Updated  tag selection and deselection functionality
- Implemented tag management with proper data structure
- Updated test mocks to match real data format
- Fix test coverage for tag management

ivilarop committed:
Fix: delete-loadRelatedChallenges-test.There are no related challenges

GuillemBo committed:
fix sonar tests

ivilarop committed:
Merge pull request #569 from IT-Academy-BCN/fix/delete-loadRelatedChallenges-test

Fix: delete-loadRelatedChallenges-test.There are no related challenges

GuillemBo committed:
Merge branch 'develop' into feature/195-isloggedIn-function

ivilarop committed:
Merge pull request #562 from IT-Academy-BCN/feature/195-isloggedIn-function

Feature/195 islogged in function

mcallejo-10 committed:
Merge feature/166: Add tag management functionality for challenges

- Implemented tag selection and display
- Added proper tag data structure
- Updated tests for tag management
- Improved UI for tag interaction merge branch 'feature/166-mentor-assign-tags-to-challenge' into develop

mcallejo-10 committed:
test: Add tags property to createChallenge mock
